src/libraries/System.Linq/src/System/Linq/Chunk.cs

```diff
@@ -50,27 +51,24 @@ public static IEnumerable<TSource[]> Chunk<TSource>(this IEnumerable<TSource> so
         private static IEnumerable<TSource[]> ChunkIterator<TSource>(IEnumerable<TSource> source, int size)
         {
             using IEnumerator<TSource> e = source.GetEnumerator();
-            while (e.MoveNext())
-            {
-                TSource[] chunk = new TSource[size];
-                chunk[0] = e.Current;
 
-                int i = 1;
-                for (; i < chunk.Length && e.MoveNext(); i++)
+            if (e.MoveNext())
+            {
+                List<TSource> chunkBuilder = new();
```

Comment:
I would recommend presizing the List up to a "reasonable" chunk size (e.g. `new(Math.Min(chunkSize, 128))`) to prevent unneeded resizes when the first chunk is filled. You could use benchmarks to determine a good cutoff number.

Reply:
This assumes there's going to be enough to fill a chunk. Do we know that's a good assumption? Benchmarks here can be used to show that any size is optimal, so how are you going to choose the benchmarks to use? To me, the reasonable sizes to use are either the chunk size (which the presence of this pr is dismissing) or the small, default value list uses internally so this method is out of the guessing game.